src/index/CompressedRelationPermutationWriterImpl.h

```diff
@@ -0,0 +1,288 @@
+    AD_LOG_TIMING << "Time spent waiting for writer2's queue "
+                  << ad_utility::Timer::toSeconds(
+                         writer2_.blockWriteQueueTimer_.msecs())
+                  << "s" << std::endl;
+    AD_LOG_TIMING << "Time spent waiting for large twin relations "
+                  << ad_utility::Timer::toSeconds(
+                         largeTwinRelationTimer_.msecs())
+                  << "s" << std::endl;
+    AD_LOG_TIMING
+        << "Time spent waiting for triple callbacks (e.g. the next sorter) "
+        << ad_utility::Timer::toSeconds(blockCallbackTimer_.msecs()) << "s"
+        << std::endl;
+  }
+
+  // ___________________________________________________________________________
+  PermutationPairResult writePermutation(
+      ad_utility::InputRangeTypeErased<IdTableStatic<0>> sortedTriples) {
+    // All columns n the order in which they have to be added to
+    // the relation.
+    auto [c0, c1, c2, c3] = permutation_.keys();
+    std::vector<ColumnIndex> permutedColIndices{c0, c1, c2};
+    for (size_t colIdx = 3; colIdx < numColumns_; ++colIdx) {
+      permutedColIndices.push_back(colIdx);
+    }
+    inputWaitTimer_.cont();
+    size_t numTriplesProcessed = 0;
+    ad_utility::ProgressBar progressBar{numTriplesProcessed,
+                                        "Triples sorted: "};
+    for (auto& block : AD_FWD(sortedTriples)) {
+      AD_CORRECTNESS_CHECK(block.numColumns() == numColumns_);
+      inputWaitTimer_.stop();
+      // This only happens when the index is completely empty.
+      if (block.empty()) {
+        continue;
+      }
+      auto firstCol = block.getColumn(c0);
+      auto permutedCols = block.asColumnSubsetView(permutedColIndices);
+      if (!col0IdCurrentRelation_.has_value()) {
+        col0IdCurrentRelation_ = firstCol[0];
+      }
+      // TODO<C++23> Use `views::zip`
+      for (size_t idx : ad_utility::integerRange(block.numRows())) {
+        Id col0Id = firstCol[idx];
+        decltype(auto) curRemainingCols = permutedCols[idx];
+
+        if (col0Id != col0IdCurrentRelation_) {
+          finishRelation();
+          col0IdCurrentRelation_ = col0Id;
+        }
+
+        // Check if we need to create a new block before adding the current
+        // triple. We create a new block if:
+        // 1. The relation buffer is at the block size limit, AND
+        // 2. The current triple has different first three columns than the last
+        //    triple in the buffer (to ensure equal triples stay in same block)
+        AD_CORRECTNESS_CHECK(blocksize_ > 0);
+        if (relation_.size() >= blocksize_) {
+          // Compare first three columns of current triple with last buffered
+          // triple
+          const auto& lastBufferedRow = relation_.back();
+          if (tieFirstThreeColumns(curRemainingCols) !=
+              tieFirstThreeColumns(lastBufferedRow)) {
+            addBlockForLargeRelation();
+          }
+        }
```

Comment:
This can be a separate function, s.t. this reads

`if (isEndOfBlockForLargeRelation(curRemainingCols, lastBufferedRow) {
addBlockForLargeRelation();
}`

Then you can have the comments at this whole function.